File: algorithm-array-processing-implementation-optimized/array_manipulation_project/src/problems/trapping_rain_water.py
def trap_rain_water_brute_force(height):
    """
    Computes the amount of water an elevation map can trap using a brute-force approach.
    For each bar, it finds the maximum height to its left and right, then calculates
    the trapped water based on the minimum of these two maximums.

    Time Complexity: O(N^2)
        - For each of the N bars, we iterate left (up to N times) and right
          (up to N times) to find the max heights.
    Space Complexity: O(1)
        - No extra space is used beyond a few variables.

    Args:
        height (list): A list of non-negative integers representing the elevation map.

    Returns:
        int: The total amount of water trapped.
    """
    if not height or len(height) < 3:
        return 0

    n = len(height)
    total_water = 0

    # We can't trap water at the very first or very last bar,
    # as there's no "wall" on both sides.
    for i in range(1, n - 1):
        # Find max height to the left of current bar (inclusive for convenience, actual limit is max_left_of_i)
        left_max = 0
        for j in range(i + 1): # Scan from 0 up to current i
            left_max = max(left_max, height[j])

        # Find max height to the right of current bar (inclusive for convenience, actual limit is max_right_of_i)
        right_max = 0
        for j in range(i, n): # Scan from current i up to n-1
            right_max = max(right_max, height[j])

        # Water trapped at current bar 'i' is limited by the minimum of
        # the max left and max right heights, minus the current bar's height.
        # We take max(0, ...) because if height[i] is greater than or equal to
        # min(left_max, right_max), no water can be trapped.
        water_at_i = min(left_max, right_max) - height[i]
        total_water += max(0, water_at_i)

    return total_water

def trap_rain_water_dp(height):
    """
    Computes the amount of water an elevation map can trap using Dynamic Programming.
    This approach precomputes the maximum height to the left and right for each position,
    then makes a single pass to calculate the trapped water.

    Time Complexity: O(N)
        - One pass to fill `left_max` array.
        - One pass to fill `right_max` array.
        - One pass to calculate total water.
    Space Complexity: O(N)
        - Two auxiliary arrays (`left_max` and `right_max`) of size N are used.

    Args:
        height (list): A list of non-negative integers representing the elevation map.

    Returns:
        int: The total amount of water trapped.
    """
    if not height or len(height) < 3:
        return 0

    n = len(height)
    left_max = [0] * n  # Stores max height from left up to current index (inclusive)
    right_max = [0] * n # Stores max height from right up to current index (inclusive)
    total_water = 0

    # Populate left_max array
    left_max[0] = height[0]
    for i in range(1, n):
        left_max[i] = max(left_max[i-1], height[i])

    # Populate right_max array
    right_max[n-1] = height[n-1]
    for i in range(n-2, -1, -1):
        right_max[i] = max(right_max[i+1], height[i])

    # Calculate trapped water for each bar
    for i in range(n):
        # The amount of water trapped above height[i] is
        # min(max_height_from_left_including_i, max_height_from_right_including_i) - height[i]
        # Since left_max[i] and right_max[i] already include height[i],
        # min(left_max[i], right_max[i]) will always be >= height[i],
        # so we don't need max(0, ...) here.
        water_at_i = min(left_max[i], right_max[i]) - height[i]
        total_water += water_at_i

    return total_water

def trap_rain_water_two_pointers(height):
    """
    Computes the amount of water an elevation map can trap using the two-pointers technique.
    This is the most optimal and space-efficient approach.

    It uses two pointers, `left` and `right`, starting from the ends of the array.
    It maintains `max_left` and `max_right` to track the maximum heights seen
    so far from the left and right, respectively.

    The key idea is that if `height[left] < height[right]`, then the water trapped
    at `left` (and potentially to its right) is limited by `max_left`. This is because
    we know there's a wall `height[right]` (or something even taller) on the right side
    that is at least as tall as `height[left]`. Thus, `max_left` is the tighter bound.
    We then move `left` inward. A symmetrical argument applies if `height[left] >= height[right]`.

    Time Complexity: O(N)
        - The two pointers traverse the array once, performing constant work at each step.
    Space Complexity: O(1)
        - No extra auxiliary arrays are used.

    Args:
        height (list): A list of non-negative integers representing the elevation map.

    Returns:
        int: The total amount of water trapped.
    """
    if not height or len(height) < 3:
        return 0

    n = len(height)
    left, right = 0, n - 1
    max_left = 0  # Max height encountered from the left up to `left`
    max_right = 0 # Max height encountered from the right up to `right`
    total_water = 0

    while left < right:
        if height[left] < height[right]:
            # The current left bar is shorter than the current right bar.
            # Water trapped at `left` (and elements between `left` and `right`)
            # will be limited by `max_left`. The right boundary is guaranteed
            # to be at least `height[right]`, which is >= height[left].
            if height[left] >= max_left:
                # This bar extends the left wall, so no water trapped at this exact position.
                max_left = height[left]
            else:
                # Water can be trapped here, limited by the existing max_left.
                total_water += max_left - height[left]
            left += 1 # Move left pointer inwards
        else:
            # The current right bar is shorter or equal to the current left bar.
            # Water trapped at `right` will be limited by `max_right`.
            # The left boundary is guaranteed to be at least `height[left]`.
            if height[right] >= max_right:
                # This bar extends the right wall.
                max_right = height[right]
            else:
                # Water can be trapped here, limited by the existing max_right.
                total_water += max_right - height[right]
            right -= 1 # Move right pointer inwards

    return total_water

# Example usage (for testing/demonstration outside unit tests)
if __name__ == "__main__":
    test_cases = [
        ([0,1,0,2,1,0,1,3,2,1,2,1], 6),
        ([4,2,0,3,2,5], 9),
        ([0,0,0,0,0], 0),
        ([1,2,3,4,5], 0), # Monotonically increasing
        ([5,4,3,2,1], 0), # Monotonically decreasing
        ([1,0,1], 1),
        ([2,0,2], 2),
        ([1,1,1,1,1], 0),
        ([], 0),
        ([1], 0),
        ([1,2], 0),
        ([4,2,3], 1)
    ]

    print("--- Trapping Rain Water (Brute Force) ---")
    for height, expected in test_cases:
        result = trap_rain_water_brute_force(list(height))
        print(f"Height: {height}, Result: {result}, Expected: {expected} {'(Correct)' if result == expected else '(Incorrect)'}")

    print("\n--- Trapping Rain Water (Dynamic Programming) ---")
    for height, expected in test_cases:
        result = trap_rain_water_dp(list(height))
        print(f"Height: {height}, Result: {result}, Expected: {expected} {'(Correct)' if result == expected else '(Incorrect)'}")

    print("\n--- Trapping Rain Water (Two Pointers) ---")
    for height, expected in test_cases:
        result = trap_rain_water_two_pointers(list(height))
        print(f"Height: {height}, Result: {result}, Expected: {expected} {'(Correct)' if result == expected else '(Incorrect)'}")

File: data-structures-array-processing-pro/src/problems/trappingRainWater.js
```javascript
/**
 * @fileoverview Problem: Trapping Rain Water
 * Given n non-negative integers representing an elevation map where the width of each bar is 1,
 * compute how much water it can trap after raining.
 *
 * Constraints:
 * n == height.length
 * 1 <= n <= 2 * 10^4
 * 0 <= height[i] <= 10^5
 */

/**
 * Approach 1: Brute Force (O(N^2)).
 * For each bar, calculate how much water it can trap.
 * The amount of water trapped above a bar at index `i` is
 * `min(max_left[i], max_right[i]) - height[i]`, if this value is positive.
 *
 * For each bar:
 * 1. Find the maximum height of a bar to its left (including itself).
 * 2. Find the maximum height of a bar to its right (including itself).
 * 3. The water it can trap is `max(0, min(max_left, max_right) - height[i])`.
 * Sum these amounts for all bars.
 *
 * @param {number[]} height - An array representing the elevation map.
 * @returns {number} The total amount of trapped rain water.
 *
 * Time Complexity: O(N^2), where N is the number of bars.
 *                  For each bar (N), we scan left (N) and scan right (N) to find max heights.
 * Space Complexity: O(1), no extra space proportional to input size.
 */
function trappingRainWater_bruteForce(height) {
    const n = height.length;
    if (n < 3) { // Need at least 3 bars to trap water
        return 0;
    }

    let totalWater = 0;

    // We cannot trap water at the very first or very last bar.
    // So, iterate from the second bar to the second-to-last bar.
    for (let i = 1; i < n - 1; i++) {
        let maxLeft = 0;
        let maxRight = 0;

        // Find max height to the left of current bar (including current bar)
        for (let j = 0; j <= i; j++) {
            maxLeft = Math.max(maxLeft, height[j]);
        }

        // Find max height to the right of current bar (including current bar)
        for (let j = i; j < n; j++) {
            maxRight = Math.max(maxRight, height[j]);
        }

        // Water trapped above current bar is min(maxLeft, maxRight) - height[i]
        // Ensure it's not negative (e.g., if height[i] is higher than both maxLeft/maxRight)
        totalWater += Math.max(0, Math.min(maxLeft, maxRight) - height[i]);
    }

    return totalWater;
}

/**
 * Approach 2: Dynamic Programming (O(N) Time, O(N) Space).
 * Precompute `max_left` and `max_right` arrays to avoid re-calculating them for each bar.
 *
 * 1. Create `leftMax` array: `leftMax[i]` stores the maximum height encountered from index 0 to `i`.
 * 2. Create `rightMax` array: `rightMax[i]` stores the maximum height encountered from index `n-1` to `i`.
 * 3. Iterate through the `height` array. For each bar `i`, the water trapped is
 *    `max(0, min(leftMax[i], rightMax[i]) - height[i])`. Sum these values.
 *
 * @param {number[]} height - An array representing the elevation map.
 * @returns {number} The total amount of trapped rain water.
 *
 * Time Complexity: O(N), for three passes over the array (one for `leftMax`, one for `rightMax`, one for calculating water).
 * Space Complexity: O(N), for storing `leftMax` and `rightMax` arrays.
 */
function trappingRainWater_dp(height) {
    const n = height.length;
    if (n < 3) {
        return 0;
    }

    const leftMax = new Array(n).fill(0);
    const rightMax = new Array(n).fill(0);

    // Populate leftMax array
    leftMax[0] = height[0];
    for (let i = 1; i < n; i++) {
        leftMax[i] = Math.max(leftMax[i - 1], height[i]);
    }

    // Populate rightMax array
    rightMax[n - 1] = height[n - 1];
    for (let i = n - 2; i >= 0; i--) {
        rightMax[i] = Math.max(rightMax[i + 1], height[i]);
    }

    let totalWater = 0;
    // Calculate trapped water
    for (let i = 0; i < n; i++) {
        totalWater += Math.max(0, Math.min(leftMax[i], rightMax[i]) - height[i]);
    }

    return totalWater;
}

/**
 * Approach 3: Two Pointers (Optimal, O(N) Time, O(1) Space).
 * This is the most efficient solution. It avoids extra space by maintaining two pointers
 * (`left` and `right`) and two variables to track `maxLeft` and `maxRight` encountered so far.
 *
 * The core idea is that if `height[left] < height[right]`:
 *  - We know the water trapped at `left` depends only on `maxLeft` (the left wall).
 *    The `right` wall is guaranteed to be at least `height[left]`, so `min(maxLeft, maxRight)`
 *    will be determined by `maxLeft`.
 *  - We increment `left`.
 *
 * Conversely, if `height[left] >= height[right]`:
 *  - The water trapped at `right` depends only on `maxRight` (the right wall).
 *    The `left` wall is guaranteed to be at least `height[right]`.
 *  - We decrement `right`.
 *
 * We update `maxLeft` and `maxRight` as we move the pointers.
 *
 * @param {number[]} height - An array representing the elevation map.
 * @returns {number} The total amount of trapped rain water.
 *
 * Time Complexity: O(N), as the two pointers traverse the array once.
 * Space Complexity: O(1), only a few variables are used.
 */
function trappingRainWater_twoPointers(height) {
    const n = height.length;
    if (n < 3) {
        return 0;
    }

    let left = 0;
    let right = n - 1;
    let maxLeft = 0;
    let maxRight = 0;
    let totalWater = 0;

    while (left < right) {
        if (height[left] < height[right]) {
            // If current left bar is smaller than current right bar,
            // we can confidently calculate water for `height[left]`.
            // The water depends on `maxLeft` (the actual left wall) and `height[left]`.
            // The right wall will always be at least `height[left]` (either current `height[right]`
            // or some future `maxRight`).
            if (height[left] >= maxLeft) {
                maxLeft = height[left]; // Update maxLeft if current bar is higher
            } else {
                totalWater += maxLeft - height[left]; // Trap water
            }
            left++; // Move left pointer
        } else {
            // If current right bar is smaller than or equal to current left bar,
            // we can confidently calculate water for `height[right]`.
            // The water depends on `maxRight` (the actual right wall) and `height[right]`.
            // The left wall will always be at least `height[right]`.
            if (height[right] >= maxRight) {
                maxRight = height[right]; // Update maxRight if current bar is higher
            } else {
                totalWater += maxRight - height[right]; // Trap water
            }
            right--; // Move right pointer
        }
    }

    return totalWater;
}

module.exports = {
    trappingRainWater_bruteForce, // Included for comparison purposes.
    trappingRainWater_dp,
    trappingRainWater_twoPointers
};
```